accepted/future-releases/records/records-feature-specification.md

```diff
@@ -257,6 +257,80 @@ avoid this ambiguity ([#2469][]).**
 
 [#2469]: https://github.com/dart-lang/language/issues/2469
 
+### Ambiguity with metadata annotations
+
+A metadata annotation may or may not have an argument list following it. A
+variable declaration may omit a preceding type annotation. Likewise, a function
+declaration may omit a preceding return type. This combination of syntax where
+an optional trailing element is followed by syntax with an optional preceding
+element can lead to ambiguity. In particular:
+
+```dart
+@metadata (a, b) function() {}
+```
+
+This could be a metadata annotation `@metadata(a, b)` associated with a function
+declaration with no return type. Or it could be a metadata annotation
+`@metadata` associated with a function whose return type is the record type `(a,
+b)`.
+
+In practice, idiomatically written code is clear thanks to whitespace:
+
+```dart
+@metadata(a, b) function() {}
+
+@metadata (a, b) function() {}
+```
+
+The former applies `(a, b)` to the metadata annotation and the latter is a
+return type. We disambiguate in the same way, by making whitespace after a
+metadata annotation name significant. Change the grammar to:
+
+```
+metadatum ::= identifier                                    // Existing rule.
+            | qualifiedName                                 // Existing rule.
+            | constructorDesignation ~WHITESPACE arguments  // Changed.
+```
+
+The `~WHITESPACE` lexical rule matches when there are no whitespace characters
+(according to the existing `WHITESPACE` lexical rule) between the
+`constructorDesignation` and `arguments`. Comments do not count as whitespace,
```

Comment:
I'd want both `~WHITESPACE` and `~COMMENT` here.

The must be *characters* between the end of the identifier and the following `(` or `<`.
That should be the easiest thing to implement (`idToken.end == nextToken.start` or something similar).

Anything more complicated is just asking for trouble.

I can see the rationale for allowing inline comments (with no surrounding whitespace), but I don't think it's actually worth it.

Reply:
SGTM.